Introduce lazyInitialize and use it in Document

rniwa opened the pull request:
#### a6c381f8c36872e7e6267099a8a53d1bf713796d
<pre>
Introduce lazyInitialize and use it in Document
<a href="https://bugs.webkit.org/show_bug.cgi?id=283038">https://bugs.webkit.org/show_bug.cgi?id=283038</a>

Reviewed by Geoffrey Garen and Chris Dumez.

In WebKit, it’s fairly common to write a member variable as `RefPtr` or `std::unique_ptr` that later
gets lazily initialized to some value but never unset or assigned of a different value after that. e.g.

```
class Foo {
    Bar&amp; bar() {
        if (!m_bar)
            m_bar = Bar::create();
        return *m_bar;
    }
    Ref&lt;Bar&gt; protectedBar() { return bar(); }

    RefPtr&lt;Bar&gt; m_bar;
}
```

Assuming there is no other code modifying m_bar, `foo-&gt;bar()-&gt;method()` is always safe to call
even if method wasn’t a trivial function. Right now, static analyzer doesn’t recognize this pattern
so we’d be forced to write code like this: `foo-&gt;protectedBar()-&gt;method()` where `protectedBar` is
a wrapper function which returns `Ref&lt;Bar&gt;`.

This PR introduces a new convention for this pattern. Namely, we would use `const Ref` or
`const UniqueRef` for when the value is initialized inside the constructor, and `const RefPtr` or
`const unique_ptr` in the case it is lazily initialized. The lazy initialization is done via newly
introduced `lazyInitialize`, which takes `const RefPtr` or `const unique_ptr` and a R-value.

e.g.
```
class Foo {
    Bar&amp; bar() {
        if (!m_bar)
            lazyInitialize(m_bar, Bar::create());
        return *m_bar;
    }

    const RefPtr&lt;Bar&gt; m_bar;
}
```

`lazyInitialize` release asserts that the pointer value isn&apos;t set and `const_cast` the pointer type
to set the value. It is a programming error to invoke `lazyInitialize` more than once on a given
instance variable in an object.

* Source/WTF/wtf/RefPtr.h:
(WTF::lazyInitialize):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::lazyInitialize):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::ensureQuirks):
(WebCore::Document::ensureExtensionStyleSheets):
(WebCore::Document::ensureMarkers):
(WebCore::Document::ensureVisitedLinkState):
(WebCore::Document::ensureFullscreenManager):
(WebCore::Document::fontLoader):
(WebCore::Document::ensureFontLoader):
(WebCore::Document::ensureFontSelector):
(WebCore::Document::ensureUndoManager):
(WebCore::Document::ensureEditor):
(WebCore::Document::ensureReportingScope):
(WebCore::Document::ensureScriptRunner):
(WebCore::Document::ensureTemplateDocument):
(WebCore::Document::ensureIntersectionObserverData):
(WebCore::Document::ensureTimelinesController):
(WebCore::Document::ensurePaintWorklet):
(WebCore::Document::ensureResizeObserverForContainIntrinsicSize):
* Source/WebCore/dom/Document.h:

Canonical link: <a href="https://commits.webkit.org/286566@main">https://commits.webkit.org/286566@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b2c50cc36638aa8951d435a74703b876e7ba287

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76353 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55385 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29255 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/80880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27634 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78469 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64527 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3679 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/80880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/18001 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79420 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49782 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65570 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/80880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47167 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23058 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/25953 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/69532 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68303 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23391 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82321 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/75629 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3725 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/2445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/68097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3879 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65542 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/67411 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11367 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/9470 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/97883 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11816 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3673 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6490 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21416 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3696 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7125 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5454 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -1091,6 +1091,13 @@ ALWAYS_INLINE constexpr void forEachSetBit(std::span<const WordType> bits, size_
 }
 
+template<typename T, typename U>
+ALWAYS_INLINE void initializeOnce(const std::unique_ptr<T>& ptr, std::unique_ptr<U>&& obj)
```
We have a bit of a name conflict with call_once and dispatch_once. In those functions, "once" promises mutual exclusion. 

Maybe we should just call this "initialize"? "Initialize" describes a starting condition, which already helps to convey that it's an error to initialize twice.

rniwa replied:
Maybe we can call it `lazilyInitialize`? `initialize` is too generic of a word and it prevents easy grep'ing.

cdumez replied:
I like lazilyInitialize or lazyInitialize, given what it's meant to be used for.

geoffreygaren replied:
I'd say lazyInitialize.

geoffreygaren reviewed (approved):
r=me

cdumez commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -1091,6 +1091,13 @@ ALWAYS_INLINE constexpr void forEachSetBit(std::span<const WordType> bits, size_
+ALWAYS_INLINE void initializeOnce(const std::unique_ptr<T>& ptr, std::unique_ptr<U>&& obj)
+{
+    RELEASE_ASSERT(!ptr);
+    const_cast<std::unique_ptr<T>&>(ptr) = std::move(obj);
```
Why aren't we using `WTFMove()` here?

rniwa replied:
Because WTFMove is defined a couple lines beneath in this file.

cdumez replied:
We can always re-order things :)

rniwa replied:
It's not as clean because of the namespace. We use `std::move` elsewhere in this file so it's probably fine. The only difference between `WTFMove` and `std::move` is that WTFMove can catch misuse but here, we know it's not an misuse.

rniwa committed:
Introduce lazyInitialize and use it in Document
https://bugs.webkit.org/show_bug.cgi?id=283038

Reviewed by Geoffrey Garen and Chris Dumez.

In WebKit, it’s fairly common to write a member variable as `RefPtr` or `std::unique_ptr` that later
gets lazily initialized to some value but never unset or assigned of a different value after that. e.g.

```
class Foo {
    Bar& bar() {
        if (!m_bar)
            m_bar = Bar::create();
        return *m_bar;
    }
    Ref<Bar> protectedBar() { return bar(); }

    RefPtr<Bar> m_bar;
}
```

Assuming there is no other code modifying m_bar, `foo->bar()->method()` is always safe to call
even if method wasn’t a trivial function. Right now, static analyzer doesn’t recognize this pattern
so we’d be forced to write code like this: `foo->protectedBar()->method()` where `protectedBar` is
a wrapper function which returns `Ref<Bar>`.

This PR introduces a new convention for this pattern. Namely, we would use `const Ref` or
`const UniqueRef` for when the value is initialized inside the constructor, and `const RefPtr` or
`const unique_ptr` in the case it is lazily initialized. The lazy initialization is done via newly
introduced `lazyInitialize`, which takes `const RefPtr` or `const unique_ptr` and a R-value.

e.g.
```
class Foo {
    Bar& bar() {
        if (!m_bar)
            lazyInitialize(m_bar, Bar::create());
        return *m_bar;
    }

    const RefPtr<Bar> m_bar;
}
```

`lazyInitialize` release asserts that the pointer value isn't set and `const_cast` the pointer type
to set the value. It is a programming error to invoke `lazyInitialize` more than once on a given
instance variable in an object.

* Source/WTF/wtf/RefPtr.h:
(WTF::lazyInitialize):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::lazyInitialize):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::ensureQuirks):
(WebCore::Document::ensureExtensionStyleSheets):
(WebCore::Document::ensureMarkers):
(WebCore::Document::ensureVisitedLinkState):
(WebCore::Document::ensureFullscreenManager):
(WebCore::Document::fontLoader):
(WebCore::Document::ensureFontLoader):
(WebCore::Document::ensureFontSelector):
(WebCore::Document::ensureUndoManager):
(WebCore::Document::ensureEditor):
(WebCore::Document::ensureReportingScope):
(WebCore::Document::ensureScriptRunner):
(WebCore::Document::ensureTemplateDocument):
(WebCore::Document::ensureIntersectionObserverData):
(WebCore::Document::ensureTimelinesController):
(WebCore::Document::ensurePaintWorklet):
(WebCore::Document::ensureResizeObserverForContainIntrinsicSize):
* Source/WebCore/dom/Document.h:

Canonical link: https://commits.webkit.org/286566@main
